File: apps/helper/Sources/OperatorDockPersistence/TaskMetadataStore.swift
import Foundation

public enum TaskState: String, Codable, Equatable, Sendable {
  case created
  case paused
  case completed
  case failed
  case cancelled
}

public struct TaskMetadataRecord: Codable, Equatable, Sendable {
  public static let currentSchemaVersion = 1

  public let schemaVersion: Int
  public let taskId: String
  public let createdAt: String
  public let state: TaskState
  public let lastEventId: String?
  public let lastCheckpointId: String?

  public init(
    schemaVersion: Int = TaskMetadataRecord.currentSchemaVersion,
    taskId: String,
    createdAt: String,
    state: TaskState,
    lastEventId: String?,
    lastCheckpointId: String?
  ) {
    self.schemaVersion = schemaVersion
    self.taskId = taskId
    self.createdAt = createdAt
    self.state = state
    self.lastEventId = lastEventId
    self.lastCheckpointId = lastCheckpointId
  }
}

public final class TaskMetadataStore: @unchecked Sendable {
  private let paths: OperatorDockPaths
  private let codec: AESGCMRecordCodec
  private let eventStore: EventStore
  private let fsyncRecorder: FsyncRecorder

  public init(
    paths: OperatorDockPaths,
    keys: PersistenceKeys,
    eventStore: EventStore,
    fsyncRecorder: FsyncRecorder = FsyncRecorder()
  ) {
    self.paths = paths
    self.codec = AESGCMRecordCodec(keys: keys)
    self.eventStore = eventStore
    self.fsyncRecorder = fsyncRecorder
  }

  public func createTask(taskId: String = UUID().uuidString) throws -> TaskMetadataRecord {
    let eventId = try eventStore.append(
      taskId: taskId,
      eventType: "task_created",
      payload: [
        "taskId": .string(taskId),
        "state": .string(TaskState.created.rawValue)
      ]
    )
    let record = TaskMetadataRecord(
      taskId: taskId,
      createdAt: iso8601Milliseconds(),
      state: .created,
      lastEventId: eventId,
      lastCheckpointId: nil
    )
    try write(record)
    return record
  }

  public func transition(taskId: String, to state: TaskState) throws -> TaskMetadataRecord {
    guard let current = try loadTask(taskId: taskId) else {
      throw LockControllerError.lockNotFound
    }

    let eventId = try eventStore.append(
      taskId: taskId,
      eventType: "task_state_transition",
      payload: [
        "from": .string(current.state.rawValue),
        "to": .string(state.rawValue)
      ]
    )
    let updated = TaskMetadataRecord(
      taskId: current.taskId,
      createdAt: current.createdAt,
      state: state,
      lastEventId: eventId,
      lastCheckpointId: current.lastCheckpointId
    )
    try write(updated)
    return updated
  }

  public func recordCheckpoint(taskId: String, checkpointEventId: String) throws -> TaskMetadataRecord {
    guard let current = try loadTask(taskId: taskId) else {
      throw LockControllerError.lockNotFound
    }

    let updated = TaskMetadataRecord(
      taskId: current.taskId,
      createdAt: current.createdAt,
      state: current.state,
      lastEventId: current.lastEventId,
      lastCheckpointId: checkpointEventId
    )
    try write(updated)
    return updated
  }

  public func loadTask(taskId: String) throws -> TaskMetadataRecord? {
    let url = metadataURL(taskId: taskId)
    guard FileManager.default.fileExists(atPath: url.path) else {
      return nil
    }
    let record = try JSONDecoder().decode(TaskMetadataRecord.self, from: codec.open(Data(contentsOf: url)))
    if record.schemaVersion > TaskMetadataRecord.currentSchemaVersion {
      throw EventStoreError.unknownFutureSchemaVersion(record.schemaVersion)
    }
    return record
  }

  private func write(_ record: TaskMetadataRecord) throws {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
    let encrypted = try codec.seal(encoder.encode(record)).bytes
    try syncWriteFile(encrypted, to: metadataURL(taskId: record.taskId), fsyncRecorder: fsyncRecorder, eventId: record.lastEventId ?? record.taskId)
  }

  private func metadataURL(taskId: String) -> URL {
    paths.tasks.appendingPathComponent("\(taskId).json")
  }
}

File: apps/daemon/src/persistence/taskMetadataStore.ts
import { existsSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import type { JsonValue } from "@operator-dock/protocol";
import { EncryptedRecordCodec } from "./encryptedRecordCodec.js";
import type { EventStore } from "./eventStore.js";
import type { PersistenceKeys } from "./persistenceKeys.js";
import type { OperatorDockPaths } from "./paths.js";

export type PersistedTaskState = "created" | "paused" | "completed" | "failed" | "cancelled";

export interface TaskMetadataRecord {
  schemaVersion: 1;
  taskId: string;
  createdAt: string;
  state: PersistedTaskState;
  lastEventId: string | null;
  lastCheckpointId: string | null;
}

export class TaskMetadataStore {
  constructor(
    private readonly paths: OperatorDockPaths,
    private readonly keys: PersistenceKeys,
    private readonly eventStore: EventStore
  ) {}

  create(taskId: string): TaskMetadataRecord {
    const eventId = this.eventStore.append(taskId, "task_created", {});
    const record: TaskMetadataRecord = {
      schemaVersion: 1,
      taskId,
      createdAt: new Date().toISOString(),
      state: "created",
      lastEventId: eventId,
      lastCheckpointId: null
    };
    this.write(record);
    return record;
  }

  transition(taskId: string, state: PersistedTaskState): TaskMetadataRecord {
    const existing = this.get(taskId);
    if (existing === undefined) {
      throw new Error(`Task metadata not found: ${taskId}`);
    }

    const eventId = this.eventStore.append(taskId, "task_state_transition", {
      from: existing.state,
      to: state
    });
    const updated = {
      ...existing,
      state,
      lastEventId: eventId
    };
    this.write(updated);
    return updated;
  }

  get(taskId: string): TaskMetadataRecord | undefined {
    const path = this.metadataPath(taskId);
    if (!existsSync(path)) {
      return undefined;
    }

    const records = EncryptedRecordCodec.readRecords<TaskMetadataRecord>(path, this.keys);
    const record = records[0]?.plaintext;
    return record?.schemaVersion === 1 ? record : undefined;
  }

  private write(record: TaskMetadataRecord): void {
    mkdirSync(this.paths.tasksRoot, { recursive: true, mode: 0o700 });
    EncryptedRecordCodec.rewriteRecords(this.metadataPath(record.taskId), this.keys, [record as unknown as Record<string, JsonValue>]);
  }

  private metadataPath(taskId: string): string {
    return join(this.paths.tasksRoot, `${taskId.replace(/[^A-Za-z0-9._-]/g, "_")}.json`);
  }
}
